### PR TITLE
Quit() to make it easy for Jobs to kill sidecar

### PR DESCRIPTION
**Pull request by tcnghia**

### CODE REVIEW

File: otel-collector/pkg/quit/quit.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+//
+// This utility function sends a POST to otel-collector sidecar's /quitquitquit endpoint.
+// Jobs can use this to terminate the sidecar upon completion.
+func Quit() {
+	var err error
+	for i := 0; i < 5; i++ {
+		if i > 1 {
+			time.Sleep(1 * time.Second)
+		}
+		_, err := http.Post("http://localhost:31415/quitquitquit", "application/json", bytes.NewBuffer(nil))
```
Comment (imjasonh) [inline otel-collector/pkg/quit/quit.go:27]:
```suggestion
		_, err := http.Post("http://localhost:31415/quitquitquit", "application/json", nil)
```

Comment (imjasonh) [inline otel-collector/pkg/quit/quit.go:27]: This shadows the outer err, so it will always be nil

File: otel-collector/pkg/quit/quit.go [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+//
+// This utility function sends a POST to otel-collector sidecar's /quitquitquit endpoint.
+// Jobs can use this to terminate the sidecar upon completion.
+func Quit() {
+	var err error
+	for i := 0; i < 5; i++ {
+		if i > 1 {
+			time.Sleep(1 * time.Second)
+		}
+		_, err := http.Post("http://localhost:31415/quitquitquit", "application/json", nil)
```
Comment (tcnghia) [inline otel-collector/pkg/quit/quit.go:27]:
```suggestion
		_, err = http.Post("http://localhost:31415/quitquitquit", "application/json", nil)
```

File: otel-collector/pkg/quit/quit.go (hunk 1 above)
Comment (tcnghia) [reply otel-collector/pkg/quit/quit.go]: Done!

Comment (tcnghia) [reply otel-collector/pkg/quit/quit.go]: Nice catch :)

File: otel-collector/pkg/quit/quit.go [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+	var err error
+	for i := 0; i < 5; i++ {
+		if i > 1 {
+			time.Sleep(1 * time.Second)
+		}
+		_, err = http.Post("http://localhost:31415/quitquitquit", "application/json", nil)
+		if err == nil {
+			log.Println("successfully POST /quitquitquit to otel-collector sidecar")
+			return
+		}
```
Comment (mattmoor) [inline otel-collector/pkg/quit/quit.go:30]: Should it check the http status code?

File: otel-collector/pkg/quit/quit.go [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+	"time"
+)
+
+// Quit explicitly manages the lifecycle of the otel-collector sidecar. When a Job is
+// injected with an otel-collector sidecar, which is an always-on container, such Job
+// will never complete.
+//
+// This utility function sends a POST to otel-collector sidecar's /quitquitquit endpoint.
+// Jobs can use this to terminate the sidecar upon completion.
+func Quit() {
```
Comment (mattmoor) [inline otel-collector/pkg/quit/quit.go:20]: Does this match the signature needed to be used with the new context OnClose (or w/e) function?

File: otel-collector/pkg/quit/quit.go [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+	var err error
+	for i := 0; i < 5; i++ {
+		if i > 1 {
+			time.Sleep(1 * time.Second)
+		}
+		_, err = http.Post("http://localhost:31415/quitquitquit", "application/json", nil)
```
Comment (mattmoor) [inline otel-collector/pkg/quit/quit.go:26]: Maybe make the port a parameter

File: otel-collector/pkg/quit/quit.go (hunk 3 above)
Comment (tcnghia) [reply otel-collector/pkg/quit/quit.go:42]: Done

File: otel-collector/pkg/quit/quit.go (hunk 4 above)
Comment (tcnghia) [reply otel-collector/pkg/quit/quit.go]: Added a test

File: otel-collector/pkg/quit/quit.go (hunk 5 above)
Comment (tcnghia) [reply otel-collector/pkg/quit/quit.go]: Done

File: otel-collector/pkg/quit/quit.go [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+		var err error
+		var resp *http.Response
+		for i := 0; i < 5; i++ {
+			if i > 1 {
+				time.Sleep(1 * time.Second)
+			}
+			resp, err := http.Post(fmt.Sprintf("http://localhost:%d/quitquitquit", port), "application/json", nil)
+			// if err is nil and resp is OK
+			if err == nil && resp.StatusCode == http.StatusOK {
+				log.Println("successfully POST /quitquitquit to otel-collector sidecar")
```
Comment (mattmoor) [inline otel-collector/pkg/quit/quit.go:36]: nit: not necessarily otel :)

File: otel-collector/pkg/quit/quit.go [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+				return
+			}
+		}
+		code := 0
+		if resp != nil {
+			code = resp.StatusCode
+		}
+		// This can happen because we don't always run the otel-collector sidecar.
+		log.Printf("cannot POST /quitquitquit to the otel-collector sidecar: err=%v, code=%d", err, code)
```
Comment (mattmoor) [inline otel-collector/pkg/quit/quit.go:45]: same nit

Comment (mattmoor) [reply otel-collector/pkg/quit/quit.go:45]:
```suggestion
		log.Printf("cannot POST /quitquitquit to the sidecar: err=%v, code=%d", err, code)
```

File: otel-collector/pkg/quit/quit.go (hunk 6 above)
Comment (mattmoor) [reply otel-collector/pkg/quit/quit.go:36]:
```suggestion
				log.Println("successfully POST /quitquitquit to sidecar")
```

**Review by mattmoor — APPROVED** [review]

**Pull request merged**